### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Flowaria**
- Event System now use cached result for comparing enemy (Better for future event interface supports)
- Enemy Glow Event now moved to interface for better reability
- ScoutScreamCustom now have missing setting (ScreamColor)
- Color Converters now support multiplier options:
```
"Color": {
    "r": 50.0
    "g": 0.0
    "b": 50.0
    "a": 50.0
}

now same as:

"Color": "#ff00ff * 50"
```
- Namespace cleanup
- Naming Convention Fixups
- EnemyDead Event Util support
- Handling Negative Damage (Healing Enemies)

### CODE REVIEW

File: ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace EECustom.CustomSettings
 {
     public static class CustomProjectileManager
     {
-        private static readonly Dictionary<byte, GameObject> _ProjectilePrefabs = new Dictionary<byte, GameObject>();
+        private static readonly Dictionary<byte, GameObject> _ProjectilePrefabs = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs:12]:
```suggestion
        private static readonly Dictionary<byte, GameObject> _projectilePrefabs = new();
```

File: ExtraEnemyCustomization/CustomSettings/CustomScoutWaveManager.cs [hunk 2]
```diff
@@ -12,11 +13,11 @@ namespace EECustom.CustomSettings
...
-        private static readonly Dictionary<string, ScoutTargetSetting> _TargetSettingDict = new Dictionary<string, ScoutTargetSetting>();
-        private static readonly Dictionary<string, ScoutWaveSetting> _WaveSettingDict = new Dictionary<string, ScoutWaveSetting>();
+        private static readonly Random _Random = new();
+        private static readonly List<ExpeditionScoutSetting> _ScoutSettings = new();
+        private static readonly List<ScoutSettingCache> _ActiveScoutSettings = new();
+        private static readonly Dictionary<string, ScoutTargetSetting> _TargetSettingDict = new();
+        private static readonly Dictionary<string, ScoutWaveSetting> _WaveSettingDict = new();
 
         private static string _PreviousExpKey = string.Empty;
         private static ExpeditionData _PreviousExpData = null;
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/CustomSettings/CustomScoutWaveManager.cs:23]:
```suggestion
        private static readonly Random _random = new();
        private static readonly List<ExpeditionScoutSetting> _scoutSettings = new();
        private static readonly List<ScoutSettingCache> _activeScoutSettings = new();
        private static readonly Dictionary<string, ScoutTargetSetting> _targetSettingDict = new();
        private static readonly Dictionary<string, ScoutWaveSetting> _waveSettingDict = new();

        private static string _previousExpKey = string.Empty;
        private static ExpeditionData _previousExpData = null;
```

File: ExtraEnemyCustomization/CustomSettings/CustomTentacleManager.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@ namespace EECustom.CustomSettings
 {
     public static class CustomTentacleManager
     {
-        private static readonly Dictionary<int, GPUC_Setup> _TentacleSetups = new Dictionary<int, GPUC_Setup>();
+        private static readonly Dictionary<int, GPUC_Setup> _TentacleSetups = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/CustomSettings/CustomTentacleManager.cs:9]:
```suggestion
        private static readonly Dictionary<int, GPUC_Setup> _tentacleSetups = new();
```

File: ExtraEnemyCustomization/Customizations/Abilities/Handlers/HealthRegenHandler.cs [hunk 4]
```diff
@@ -15,14 +16,14 @@ public class HealthRegenHandler : MonoBehaviour
...
+        private float _regenInitialTimer = 0.0f;
+        private float _regenIntervalTimer = 0.0f;
+        private bool _isRegening = false;
+        private bool _isInitialTimerDone = false;
+        private bool _alwaysRegen = false;
+        private bool _isDecay = false;
+        private float _regenCapAbsValue = 0.0f;
+        private float _regenAmountAbsValue = 0.0f;
 
         private Action<EnemyAgent, Agent> _OnDamageDel;
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/Abilities/Handlers/HealthRegenHandler.cs:28]: Inconsistent naming with the rest of the members
```suggestion
        private Action<EnemyAgent, Agent> _onDamageDel;
```

File: ExtraEnemyCustomization/Customizations/EnemyCustomBase.cs [hunk 5]
```diff
@@ -13,12 +15,38 @@ public abstract class EnemyCustomBase
         public bool Enabled { get; set; } = true;
         public TargetSetting Target { get; set; } = new TargetSetting();
 
+        private readonly Dictionary<ushort, bool> _IsTargetLookup = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/EnemyCustomBase.cs:18]:
```suggestion
        private readonly Dictionary<ushort, bool> _isTargetLookup = new();
```

File: ExtraEnemyCustomization/Customizations/Models/GlowCustom.cs [hunk 6]
```diff
@@ -1,32 +1,28 @@
...
 
 namespace EECustom.Customizations.Models
 {
-    public class GlowCustom : EnemyCustomBase, IEnemySpawnedEvent
+    public class GlowCustom : EnemyCustomBase, IEnemySpawnedEvent, IEnemyGlowEvent
     {
         private readonly static Color DefaultPropaWakeColor = ES_HibernateWakeUp.m_propagatedWakeupColor;
-        private readonly static Vector4 DefaultPropaWakeLocation = ES_HibernateWakeUp.m_propagatedWakeupLocation;
-        private readonly static System.Random _Random = new System.Random();
+        private readonly static System.Random _Random = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/Models/GlowCustom.cs:13]:
```suggestion
        private readonly static System.Random _random = new();
```

File: ExtraEnemyCustomization/Customizations/Models/MaterialCustom.cs [hunk 7]
```diff
@@ -7,7 +7,7 @@ namespace EECustom.Customizations.Models
 {
     public class MaterialCustom : EnemyCustomBase, IEnemyPrefabBuiltEvent
     {
-        private readonly static Dictionary<string, Material> _MatDict = new Dictionary<string, Material>();
+        private readonly static Dictionary<string, Material> _MatDict = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/Models/MaterialCustom.cs:10]:
```suggestion
        private readonly static Dictionary<string, Material> _matDict = new();
```

File: ExtraEnemyCustomization/Managers/ConfigManager__Events.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                        if (!IgnoreLogs) custom.LogVerbose($"Finished!");
+                    }
+                }
+            }
+        }
+
+        private readonly EventHolder<IEnemyPrefabBuiltEvent> _EnemyPrefabBuiltHolder = new("PrefabBuilt");
+        private readonly EventHolder<IEnemySpawnedEvent> _EnemySpawnedHolder = new("Spawned");
+        private readonly EventHolder<IEnemyDespawnedEvent> _EnemyDespawnedHolder = new("Despawned");
+        private readonly EventHolder<IEnemyGlowEvent> _EnemyGlowHolder = new("Glow", true);
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Managers/ConfigManager__Events.cs:100]:
```suggestion
        private readonly EventHolder<IEnemyPrefabBuiltEvent> _enemyPrefabBuiltHolder = new("PrefabBuilt");
        private readonly EventHolder<IEnemySpawnedEvent> _enemySpawnedHolder = new("Spawned");
        private readonly EventHolder<IEnemyDespawnedEvent> _enemyDespawnedHolder = new("Despawned");
        private readonly EventHolder<IEnemyGlowEvent> _enemyGlowHolder = new("Glow", true);
```

File: ExtraEnemyCustomization/Managers/SpriteManager.cs [hunk 9]
```diff
@@ -8,14 +8,14 @@ public class SpriteManager
     {
         public static string BaseSpritePath { get; private set; }
 
-        private static readonly Dictionary<string, Texture2D> _TextureCache = new Dictionary<string, Texture2D>();
-        private static readonly Dictionary<string, Sprite> _SpriteCache = new Dictionary<string, Sprite>();
+        private static readonly Dictionary<string, Texture2D> _TextureCache = new();
+        private static readonly Dictionary<string, Sprite> _SpriteCache = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Managers/SpriteManager.cs:12]:
```suggestion
        private static readonly Dictionary<string, Texture2D> _textureCache = new();
        private static readonly Dictionary<string, Sprite> _spriteCache = new();
```

File: ExtraEnemyCustomization/Utils/EnemyProperty.cs [hunk 10]
```diff
@@ -1,21 +1,20 @@
...
-using System;
 using System.Collections.Generic;
-using System.Text;
 
 namespace EECustom.Utils
 {
     public static class EnemyProperty<T> where T : class, new()
     {
-        private readonly static Dictionary<ushort, T> _Properties = new Dictionary<ushort, T>();
+        private readonly static Dictionary<ushort, T> _Properties = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Utils/EnemyProperty.cs:10]:
```suggestion
        private readonly static Dictionary<ushort, T> _properties = new();
```

File: ExtraEnemyCustomization/Utils/JSON.cs [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
+using EECustom.Utils.JsonConverters;
+using System.Text.Json;
 using System.Text.Json.Serialization;
 
 namespace EECustom.Utils
 {
     public static class JSON
     {
-        private readonly static JsonSerializerOptions _Setting = new JsonSerializerOptions()
+        private readonly static JsonSerializerOptions _Setting = new()
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Utils/JSON.cs:10]:
```suggestion
        private readonly static JsonSerializerOptions _setting = new()
```

File: ExtraEnemyCustomization/Utils/RegexUtil.cs [hunk 12]
```diff
@@ -1,13 +1,11 @@
...
-using System.Collections.Generic;
-using System.Text;
 using System.Text.RegularExpressions;
 
 namespace EECustom.Utils
 {
     public static class RegexUtil
     {
-        public static readonly Regex VectorRegex = new Regex("[0-9.]+"); 
+        public static readonly Regex VectorRegex = new("-?[0-9.]+");
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Utils/RegexUtil.cs:8]: Might prefer to keep this regex private? I don't see a use for it being public

**Review by Kasuromi — CHANGES_REQUESTED** [review]
Minor naming inconsistencies, some I couldn't review because the diff didn't show them. Other than those everything looks good as is.

**Comment by Panthr75** [comment]
Why kas. Why all the new()

File: ExtraEnemyCustomization/Utils/RegexUtil.cs (hunk 12 above)
Comment (Flowaria) [reply ExtraEnemyCustomization/Utils/RegexUtil.cs]: Fair enough, I'll move it to private

**Comment by Flowaria** [comment]
> 
> 
> Why kas. Why all the new()

Why not?

**Review by Kasuromi — CHANGES_REQUESTED** [review]
Minor naming inconsistencies

**Review by Kasuromi — APPROVED** [review]
Looks good to me

**Pull request merged**